html5ever/src/tree_builder/rules.rs

```diff
@@ -34,19 +33,74 @@ fn current_node<Handle>(open_elems: &[Handle]) -> &Handle {
     open_elems.last().expect("no current element")
 }
 
+/// Helper macro that generates a [pattern](https://doc.rust-lang.org/reference/patterns.html) representing
+/// a [`Tag`] to make matching on [`Tag`]s less verbose.
+///
+/// This macro accepts 4 forms:
+///
+/// - `tag!(<div>)` where `div` can be any valid tag name. This matches a start tag where the tag name is "div".
+///   If the tag name contains characters other than [a-zA-Z0-9_] then it should be quoted a `<"div">`.
+/// - `tag!(</div>)` where `div` can be any valid tag name. This matches a end tag where the tag name is "div".
+///   If the tag name contains characters other than [a-zA-Z0-9_] then it should be quoted a `</"div">`.
+/// - `tag!(<>)`. This matches any start tag (regardless of tag name).
+/// - `tag!(</>)`. This matches any end tag (regardless of tag name).
+///
+/// Additionally any of the above can be freely combined with `|` to create an "or" match pattern.
+/// For example `tag!(<head> | </>)` will match a "head" start tag or any end tag.
+#[rustfmt::skip]
+macro_rules! tag {
+    // Any start tag
+    (<>) => {
+        crate::tokenizer::Tag { kind: crate::tokenizer::StartTag, .. }
+    };
+    (<>|$($tail:tt)*) => {
+        tag!(<>) | tag!($($tail)*)
+    };
+
+    // Any end tag
+    (</>) => {
+        crate::tokenizer::Tag { kind: crate::tokenizer::EndTag, .. }
+    };
+    (</>|$($tail:tt)*) => {
+        tag!(</>) | tag!($($tail)*)
+    };
+
+    // Named start tag
+    (<$tag:tt>) => {
+        crate::tokenizer::Tag { kind: crate::tokenizer::StartTag, name: local_name!($tag), .. }
+    };
+    (<$tag:tt>|$($tail:tt)*) => {
+        tag!(<$tag>) | tag!($($tail)*)
+    };
+
+    // Named end tag
+    (</$tag:tt>) => {
+        crate::tokenizer::Tag { kind: crate::tokenizer::EndTag, name: local_name!($tag), .. }
+    };
+    (</$tag:tt>|$($tail:tt)*) => {
+        tag!(</$tag>) | tag!($($tail)*)
+    };
+}
+
 #[doc(hidden)]
 impl<Handle, Sink> TreeBuilder<Handle, Sink>
 where
     Handle: Clone,
     Sink: TreeSink<Handle = Handle>,
 {
+    /// Process an HTML content token
+    ///
+    /// <https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inhtml>
     pub(crate) fn step(&self, mode: InsertionMode, token: Token) -> ProcessResult<Handle> {
         self.debug_step(mode, &token);
 
         match mode {
-            //§ the-initial-insertion-mode
-            InsertionMode::Initial => match_token!(token {
-                Token::Characters(SplitStatus::NotSplit, text) => ProcessResult::SplitWhitespace(text),
+            // § The "initial" insertion mode
+            // <https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode>
```

Comment:
It's a bit redundant to have both the spec link and the paragraph title. Also this is a regular comment, so the `<>` around the link are not required.

This is a small nit, feel free to keep it this way if it's too much effort to change all these references again :D

Reply:
I've removed the `<>` (easy with multiple cursors), but I've left the text labels as some of them differ from the link fragment (e.g. `§ The "text" insertion mode` and `https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-incdata`)